File: lib/data/distance_matrix_sync.dart
import '../domain/models/coordinates.dart';
import '../domain/models/distance_matrix_entry.dart';
import '../infra/services/ors_routing_service.dart';
import 'repositories/client_repository.dart';
import 'repositories/distance_matrix_repository.dart';
import 'repositories/settings_repository.dart';

class DistanceMatrixSync {
  final ClientRepository clients;
  final DistanceMatrixRepository matrix;
  final SettingsRepository settings;
  final OrsRoutingService ors;

  DistanceMatrixSync({
    required this.clients,
    required this.matrix,
    required this.settings,
    required this.ors,
  });

  /// Compute outbound (X→all) and inbound (all→X) rows for [clientId],
  /// where "all" = base + every other existing client.
  Future<void> recomputeForClient(int clientId) async {
    final settingsRow = await settings.read();
    if (settingsRow == null) {
      throw StateError('Cannot compute matrix without home base');
    }
    final target = await clients.findById(clientId);
    if (target == null) {
      throw StateError('Unknown client $clientId');
    }
    final others = (await clients.listAll())
        .where((c) => c.id != clientId)
        .toList();

    // Build the locations list: index 0 = X, then base, then every other.
    final locations = <Coordinates>[
      target.coordinates,
      settingsRow.baseCoordinates,
      ...others.map((c) => c.coordinates),
    ];
    final ids = <int>[
      clientId,
      DistanceMatrixEntry.baseId,
      ...others.map((c) => c.id),
    ];

    try {
      // Outbound: from X (index 0) to all (indices 1..n)
      final outbound = await ors.matrix(
        locations: locations,
        sources: const [0],
        destinations: List.generate(locations.length - 1, (i) => i + 1),
      );
      // Inbound: from all (indices 1..n) to X (index 0)
      final inbound = await ors.matrix(
        locations: locations,
        sources: List.generate(locations.length - 1, (i) => i + 1),
        destinations: const [0],
      );

      final now = DateTime.now();
      final rows = <DistanceMatrixEntry>[];
      for (var j = 0; j < ids.length - 1; j++) {
        rows.add(DistanceMatrixEntry(
          fromId: ids[0],
          toId: ids[j + 1],
          distanceMeters: outbound.distances[0][j],
          durationSeconds: outbound.durations[0][j],
          computedAt: now,
        ));
        rows.add(DistanceMatrixEntry(
          fromId: ids[j + 1],
          toId: ids[0],
          distanceMeters: inbound.distances[j][0],
          durationSeconds: inbound.durations[j][0],
          computedAt: now,
        ));
      }
      await matrix.upsertMany(rows);
      await clients.setRecomputeDone(clientId);
    } on OrsException {
      await clients.setRecomputePending(clientId);
      rethrow;
    }
  }

  /// Recompute base ↔ all clients (when the base address changes).
  Future<void> recomputeAllForBase() async {
    final settingsRow = await settings.read();
    if (settingsRow == null) return;
    final all = await clients.listAll();
    if (all.isEmpty) return;

    final locations = <Coordinates>[
      settingsRow.baseCoordinates,
      ...all.map((c) => c.coordinates),
    ];
    final ids = <int>[
      DistanceMatrixEntry.baseId,
      ...all.map((c) => c.id),
    ];

    final outbound = await ors.matrix(
      locations: locations,
      sources: const [0],
      destinations: List.generate(all.length, (i) => i + 1),
    );
    final inbound = await ors.matrix(
      locations: locations,
      sources: List.generate(all.length, (i) => i + 1),
      destinations: const [0],
    );

    final now = DateTime.now();
    final rows = <DistanceMatrixEntry>[];
    for (var j = 0; j < all.length; j++) {
      rows.add(DistanceMatrixEntry(
        fromId: ids[0],
        toId: ids[j + 1],
        distanceMeters: outbound.distances[0][j],
        durationSeconds: outbound.durations[0][j],
        computedAt: now,
      ));
      rows.add(DistanceMatrixEntry(
        fromId: ids[j + 1],
        toId: ids[0],
        distanceMeters: inbound.distances[j][0],
        durationSeconds: inbound.durations[j][0],
        computedAt: now,
      ));
    }
    await matrix.upsertMany(rows);
  }

  /// Retry every client flagged `needsDistanceRecompute`.
  Future<int> retryAllPending() async {
    final pending = await clients.listNeedingRecompute();
    var done = 0;
    for (final c in pending) {
      try {
        await recomputeForClient(c.id);
        done++;
      } on OrsException {
        // Leave it pending; banner stays.
      }
    }
    return done;
  }
}

File: src/data/distance-matrix-sync.ts
import type { Db } from '@/infra/db/client';
import { ClientRepository } from '@/data/repositories/client-repository';
import { DistanceMatrixRepository } from '@/data/repositories/distance-matrix-repository';
import { SettingsRepository } from '@/data/repositories/settings-repository';
import { fetchPairFromBase } from '@/infra/services/ors-routing';

export class DistanceMatrixSync {
  private clients: ClientRepository;
  private matrix: DistanceMatrixRepository;
  private settings: SettingsRepository;

  constructor(db: Db) {
    this.clients = new ClientRepository(db);
    this.matrix = new DistanceMatrixRepository(db);
    this.settings = new SettingsRepository(db);
  }

  /**
   * Recompute the BASE↔client pair for a single client.
   * Marks the client as recompute-pending if the fetch fails;
   * clears the flag and writes both rows on success.
   */
  async recomputeForClient(clientId: string): Promise<{ ok: boolean; reason?: string }> {
    const client = await this.clients.byId(clientId);
    if (!client) return { ok: false, reason: 'client not found' };
    if (client.latitude == null || client.longitude == null) {
      return { ok: false, reason: 'client has no coordinates' };
    }
    const baseLat = await this.settings.get('base_lat');
    const baseLng = await this.settings.get('base_lng');
    if (!baseLat || !baseLng) return { ok: false, reason: 'no base configured' };
    const base = { lat: parseFloat(baseLat), lon: parseFloat(baseLng) };
    const now = new Date().toISOString();

    try {
      const { forward, reverse } = await fetchPairFromBase(base, {
        lat: client.latitude,
        lon: client.longitude,
      });
      await this.matrix.upsert({
        fromId: 'BASE',
        toId: clientId,
        distanceKm: forward.distanceKm,
        durationMinutes: forward.durationMinutes,
        fetchedAt: now,
        failed: false,
      });
      await this.matrix.upsert({
        fromId: clientId,
        toId: 'BASE',
        distanceKm: reverse.distanceKm,
        durationMinutes: reverse.durationMinutes,
        fetchedAt: now,
        failed: false,
      });
      await this.clients.setRecomputePending(clientId, false, now);
      return { ok: true };
    } catch (err) {
      await this.matrix.markFailed('BASE', clientId, now);
      await this.matrix.markFailed(clientId, 'BASE', now);
      // Keep flag set so the banner stays
      await this.clients.setRecomputePending(clientId, true, now);
      return { ok: false, reason: err instanceof Error ? err.message : 'unknown' };
    }
  }

  /**
   * Recompute BASE↔client pairs for ALL geocoded clients.
   * Used after the base address changes.
   * Returns counts; doesn't throw on individual failures.
   */
  async recomputeAllForBase(): Promise<{ ok: number; failed: number; skipped: number }> {
    const allClients = await this.clients.listAll();
    let ok = 0;
    let failed = 0;
    let skipped = 0;
    for (const c of allClients) {
      if (c.latitude == null || c.longitude == null) {
        skipped++;
        continue;
      }
      const result = await this.recomputeForClient(c.id);
      if (result.ok) ok++;
      else failed++;
    }
    return { ok, failed, skipped };
  }

  /**
   * Mark every geocoded client as recompute-pending without fetching.
   * Used when the base changes — the user can then trigger the recompute
   * from the banner CTA, avoiding a long blocking call.
   */
  async markAllPending(): Promise<number> {
    const allClients = await this.clients.listAll();
    const now = new Date().toISOString();
    let count = 0;
    for (const c of allClients) {
      if (c.latitude == null || c.longitude == null) continue;
      await this.clients.setRecomputePending(c.id, true, now);
      count++;
    }
    return count;
  }
}
